### PR TITLE
User/jens varughese Advanced Recruit Training

### PR DESCRIPTION
**Pull request by JensVarughese**
I Finish Advanced Recruit Training

### CODE REVIEW

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 1]
```diff
@@ -16,17 +17,41 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    int seconds;
+
+    if(isCharging())
+        totalHours = -1 * totalAmpHoursUsed() / current_;
+    else
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY-totalAmpHoursUsed()) / current_;
+
+    hours = (int)totalHours;
+    minutes = (int)((totalHours - hours) * 60);
+    seconds = (int)((totalHours*60 - hours*60 - minutes) * 60);
```
Comment (JaydenBischoff) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Include spacing before and after the multiplication operation.

**Comment by bill-luu** [comment]
I'll accept the answer if the textstream output is:
`Time until fully charged: "HH:MM:SS.ZZZ"`

So you will have to put milliseconds into your QTime.

I would suggest storing your overflow into a new variable!

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 2]
```diff
@@ -1,12 +1,17 @@
 #include "BatteryStateOfChargeService.h"
+#include "BatteryData.h"
+#include "QTime"
 
 namespace
 {
-    const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
+const int SECONDS_TO_HOURS = 3600;
+const int PERCENT_TO_DECMIMAL = 100;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Misspelled decimal

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 3]
```diff
@@ -16,23 +21,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(timeOld.isNull()) //checks if timeOld has no a value (first itteration)
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: `no a value` is not english, also misspelled iteration.

This kinda just reiterates the line same line, explain why you are checking this.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 4]
```diff
@@ -29,4 +30,8 @@ class BatteryStateOfChargeService : public I_BatteryStateOfChargeService
     const double initialStateOfChargePercent_;
 
     // Add your necessary variables here
+    //double current_;
+    double currentNew_;
+    double ampHours_ = 0;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: Don't initialize this here.

File: Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 5]
```diff
@@ -2,14 +2,15 @@
 #include <QFile>
 #include <QString>
 #include <QTextStream>
+#include <exception>
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp]: Unused?

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 6]
```diff
@@ -21,12 +22,20 @@ class BatteryStateOfChargeService : public I_BatteryStateOfChargeService
      * Return the time remaining to when the battery is charged or depleted*/
     QTime timeWhenChargedOrDepleted() const;
 
+    int getRemainingHours() const;
+
     /*Update Variables*/
     void addData(const BatteryData& batteryData);
+    void setRemainingHours(int h);
```
Comment (navjotbrar) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: Change "h" to something more descriptive.

Comment (JensVarughese) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: setRemainingHours was removed in a later commit

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp [hunk 7]
```diff
@@ -26,8 +26,20 @@ void BatteryStateDisplayService::handleBatteryDataReceived(const BatteryData& ba
...
-        << " Current: " << batteryData.current
-        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed() << endl;
+                        << " Current: " << batteryData.current
+                        << " Total Ah used: " << batteryStateOfChargeService_.totalAmpHoursUsed()<< endl;
 
     // TODO: Print out time until the battery is fully charged or depleted.
+    QTime displayTime = batteryStateOfChargeService_.timeWhenChargedOrDepleted();
+
+    if(batteryStateOfChargeService_.isCharging())
+        QTextStream(stdout)<<"Time until charge:";
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: Leave a space between operators:
`QTextStream(stdout) << "Time...`

Comment (bill-luu) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: Also, technically it's "Time until _charged_" :p

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp [hunk 8]
```diff
@@ -26,8 +26,20 @@ void BatteryStateDisplayService::handleBatteryDataReceived(const BatteryData& ba
...
+    else
+        QTextStream(stdout)<<"Time until depletion:";
+
+    QTextStream(stdout) <<"  Hours:" << (displayTime.hour() + batteryStateOfChargeService_.getRemainingHours())
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: Should be in the format `HH:MM:SS.ZZZ`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 9]
```diff
@@ -1,12 +1,18 @@
 #include "BatteryStateOfChargeService.h"
+#include "BatteryData.h"
+#include "QTime"
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Library headers should be in angled braces. Quotes are usually reserved for files in the project.
`<QTime>`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 10]
```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return ampHours_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(currentNew_ < 0)
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: You can also just do

`return (currentNew_ < 0)`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 11]
```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+    QTime timeOld = timeNew_;
+    double currentOld = currentNew_;
+
+    currentNew_ = batteryData.current;
+    timeNew_ = batteryData.time;
+
+    if(timeOld.isNull()) // first itteration
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: iteration is misspelled

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 12]
```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    {
+        ampHours_ = BATTERY_AMP_HOUR_CAPACITY * initialStateOfChargePercent_ / PERCENT_TO_DECIMAL;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: This looks like this should be updated in the initializer list, since it's starting value of `ampHours_`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h [hunk 13]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "I_BatteryStateOfChargeService.h"
+#include <QTime>
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.h]: Do you need this? Your .cpp already has this include.

If you do, you should forward declare it instead.
https://stackoverflow.com/questions/4757565/what-are-forward-declarations-in-c

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 14]
```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    if(isCharging())
+    {
+        totalHours = qAbs(totalAmpHoursUsed() / currentNew_);
+    }
+    else
+    {
+        totalHours = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / currentNew_;
+    }
+
+    hours = (int)totalHours;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: These calculations are pretty hard to follow, especially with all the repeated arithmetic.

Try converting your total time to milliseconds, and work from there, similar to this:
https://stackoverflow.com/questions/9102680/how-to-convert-a-decimal-into-time-eg-hhmmss

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 15]
```diff
@@ -16,23 +22,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+    minutes = (int)((totalHours - hours) * HMS_CONVERSION_FACTOR);
+    seconds = (int)((totalHours * HMS_CONVERSION_FACTOR - hours * HMS_CONVERSION_FACTOR - minutes) * HMS_CONVERSION_FACTOR);
+
+    remainingHours_ = ((int)(totalHours / HOURS_IN_A_DAY)) * HOURS_IN_A_DAY;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: This comment will be miscellaneous from the above, but it looks like you could have used `remainingHours_ = hours % HOURS_IN_A_DAY`?

Comment (JensVarughese) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: remainingHours_ can be larger than 24

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp [hunk 16]
```diff
@@ -26,8 +26,20 @@ void BatteryStateDisplayService::handleBatteryDataReceived(const BatteryData& ba
...
+    if(batteryStateOfChargeService_.isCharging())
+        QTextStream(stdout) << "Time until charged:";
+    else
+        QTextStream(stdout) << "Time until depleted:";
+
+    QTextStream(stdout) << "  Hours:" << (displayTime.hour() + batteryStateOfChargeService_.getRemainingHours())
+                        << "  Min:" << batteryStateOfChargeService_.timeWhenChargedOrDepleted().minute()
+                        << "  Sec:" << batteryStateOfChargeService_.timeWhenChargedOrDepleted().second()
+                        << "  Ms:" << batteryStateOfChargeService_.timeWhenChargedOrDepleted().msec()
+                        << "\n" << endl;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: `endl` should already print a new line, so `\n` isn't needed.

Comment (bill-luu) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateDisplayService.cpp]: Also, your format is `Hours: <HH> Min: <MM> Sec: <SS> Ms: <ZZZ>`
It should be `HH:MM:SS.ZZ`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 17]
```diff
@@ -16,23 +25,75 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+
+    minutes = floor(milliseconds / MILLISECONDS_IN_MINUTE);
+    milliseconds -= minutes * MILLISECONDS_IN_MINUTE;
+
+    seconds = floor(milliseconds / MILLISECONDS_IN_SECOND);
+    milliseconds -= seconds * MILLISECONDS_IN_SECOND;
+
+    if((hours - HOURS_IN_A_DAY) >= 0) // if 24 hours or more
+    {
+        remainingHours_ = hours - (HOURS_IN_A_DAY - 1);
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: Why -1?
If you're at 30 total hours, shouldn't the remainingHours_ be 6? Instead of 30 - (24 - 1) = 7?

Comment (JensVarughese) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: The hours parameter in QTime will only hold up to 23 hours, not 24.

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 18]
```diff
@@ -16,23 +25,75 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+        double avgCurrent = (currentOld + currentNew_) / 2;
+        ampHours_ += (avgCurrent * timeOld.secsTo(timeNew_) / SECONDS_TO_HOURS);
+    }
+
+    //calculate remaining time for charge or depletion
+    double totalHours;
+    int hours;
+    int minutes;
+    int seconds;
+    int milliseconds;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: These aren't used until after the if statement, so we can move it down.

I think we can also just declare and set these variables, ie instead of 
```
int hours
...
hours = ....
```
Just do
```
int hours = ...
```

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp (hunk 17 above)
Comment (bill-luu) [reply Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: I think this line deserves a comment then

File: Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 19]
```diff
@@ -7,9 +7,9 @@
 
 namespace
 {
-    const QString STRING_TIME_FORMAT= "hh:mm:ss.zzz";
-    const QString BATDATA_DELIMITER= ", ";
-    const int COLUMNS = 3;
+const QString STRING_TIME_FORMAT= "hh:mm:ss.zzz";
+const QString BATDATA_DELIMITER= ", ";
+const int COLUMNS = 3;
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp]: Random whitespace change?

File: Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 20]
```diff
@@ -58,13 +58,24 @@ bool LogFileReader::readAll(const QString& fileName)
...
+    bool currentOk;
+
+    batteryData.voltage = sections.at(1).toDouble(&voltageOk);
+
+    batteryData.current = sections.at(2).toDouble(&currentOk);
+
+    if(!batteryData.time.isValid() || !voltageOk || !currentOk)
+    {
+        return false;
+    }
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp]: You can also just `return  (batteryData.time.isValid() && voltageOk && currentOk)`

File: Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp [hunk 21]
```diff
@@ -16,23 +25,71 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
...
+
+    int minutes = floor(milliseconds / MILLISECONDS_IN_MINUTE);
+    milliseconds -= minutes * MILLISECONDS_IN_MINUTE;
+
+    int seconds = floor(milliseconds / MILLISECONDS_IN_SECOND);
+    milliseconds -= seconds * MILLISECONDS_IN_SECOND;
+
+    if((hours - HOURS_IN_A_DAY) >= 0) // if 24 hours or more
+    {
+        remainingHours_ = hours - (HOURS_IN_A_DAY - 1); // calculate the amount of hours needed to take out of "hours" as "hours" can only hold up to 23
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp]: This line is getting too long, let's put the comment the line above. No strict rule on line length, but a good rule of thumb that if it can't fit when you do split screen on your editor, it's too long. 

Also I think we can condense this comment to just: `QTime can only hold up to 23 hours` 
Unless the code isn't straightforward enough to just read, most comments should be supplementary - explaining the intention behind the change, instead of what the change does.

File: Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp [hunk 22]
```diff
@@ -37,7 +37,7 @@ bool LogFileReader::readAll(const QString& fileName)
         if (!parseLine(line, batteryData))
         {
             qDebug() << "Error while parsing" << line;
-            // return false;
+
```
Comment (bill-luu) [inline Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp]: Extra whitespace

**Comment by bill-luu** [comment]
image
